controllers/clustergroupupgrade_controller.go

```diff
@@ -973,14 +973,93 @@ func (r *ClusterGroupUpgradeReconciler) reconcileResources(ctx context.Context,
+	// appear as being NonCompliant.
+	clustersNonCompliant := make(map[string]int)
+	for _, clusterName := range clusterGroupUpgrade.Spec.Clusters {
+		clustersNonCompliant[clusterName] = 0
+	}
+
+	for _, managedPolicy := range managedPolicies {
+		statusObject := managedPolicy.Object["status"].(map[string]interface{})
+		statusCompliance := statusObject["status"]
+
+		if statusCompliance == nil {
+			continue
+		}
+
+		for key, value := range statusObject {
+			r.Log.Info("key and value", "key", key, "value", value)
+		}
+
+		subStatus := statusCompliance.([]interface{})
+
+		for subStatusIndex := range subStatus {
+			crtSubStatus := subStatus[subStatusIndex].(map[string]interface{})
+			// If the cluster is NonCompliant in this current policy and it's also present in the list of clusters from the CR
+			// increment the number of policies for which it's NonCompliant.
+			if crtSubStatus["compliant"] == "NonCompliant" {
+				if _, ok := clustersNonCompliant[crtSubStatus["clustername"].(string)]; ok {
+					clustersNonCompliant[crtSubStatus["clustername"].(string)]++
+				}
+			}
+		}
+	}
+
+	for clusterName, clusterNonCompliantNumber := range clustersNonCompliant {
+		if clusterNonCompliantNumber == len(managedPolicies) {
+			clusterMap[clusterName] = true
+		} else {
+			clusterMap[clusterName] = false
+		}
+	}
+	r.Log.Info("Cluster (non) compliant with all managedPolicies map  : ", "clustersNotCompliantMap", clusterMap)
+	return clusterMap
+}
+
+func (r *ClusterGroupUpgradeReconciler) buildRemediationPlan(ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade, managedPolicies []*unstructured.Unstructured) {
+	clusterNonCompliantWithManagedPoliciesMap := r.getClustersNonCompliantWithAllManagedPolicies(ctx, clusterGroupUpgrade, managedPolicies)
+
+	/*  TODO: remove this at cleanup, keep it just for testing locally with kind.
+	    clusterNonCompliantWithManagedPoliciesMap := make(map[string]bool)
+		clusterNonCompliantWithManagedPoliciesMap["spoke1"] = true
+		clusterNonCompliantWithManagedPoliciesMap["spoke2"] = true
+		clusterNonCompliantWithManagedPoliciesMap["spoke3"] = true
+		clusterNonCompliantWithManagedPoliciesMap["spoke4"] = true
+		clusterNonCompliantWithManagedPoliciesMap["spoke5"] = true
+		clusterNonCompliantWithManagedPoliciesMap["spoke6"] = true
+	*/
+
 	// Create remediation plan
 	var remediationPlan [][]string
 	isCanary := make(map[string]bool)
 	if clusterGroupUpgrade.Spec.RemediationStrategy.Canaries != nil && len(clusterGroupUpgrade.Spec.RemediationStrategy.Canaries) > 0 {
 		for _, canary := range clusterGroupUpgrade.Spec.RemediationStrategy.Canaries {
-			remediationPlan = append(remediationPlan, []string{canary})
-			isCanary[canary] = true
+			// TODO: make sure the canary clusters are in the list of clusters.
+			if clusterNonCompliantWithManagedPoliciesMap[canary] == true {
+				remediationPlan = append(remediationPlan, []string{canary})
+				isCanary[canary] = true
+			}
 		}
 
 	}
```

Comment:
Kind of strange logic or I miss understand; what make a cluster exist in both `clusterGroupUpgrade.Spec.RemediationStrategy.Canaries` AND `clusterGroupUpgrade.Spec.Clusters` why there are 2 for loops to identify if the cluster is canary ? if it is a canary cluster so it will be in the canary list if not then it will be in the cluster list.

Reply:
This is actually the initial implementation which contained all the clusters that we want to update in the `clusterGroupUpgrade.Spec.Clusters` and if from that list we want any of them to be canaries, we'll put them in `clusterGroupUpgrade.Spec.RemediationStrategy.Canaries`. I haven't changed that initial logic.